phyplus_gatt.c

```diff
@@ -0,0 +1,770 @@
+ *          MI_ERR_INVALID_STATE   Invalid Connection State or notifications
+ * and/or indications not enabled in the CCCD.
+ *          MI_ERR_INVALID_LENGTH   Invalid length supplied.
+ *          MI_ERR_BUSY            Procedure already in progress.
+ *          MIBLE_ERR_ATT_INVALID_HANDLE     Attribute not found.
+ *          MIBLE_ERR_GATT_INVALID_ATT_TYPE   //Attributes are not modifiable by
+ * the application.
+ * @note    This function checks for the relevant Client Characteristic
+ * Configuration descriptor value to verify that the relevant operation (notification or
+ * indication) has been enabled by the client.
+ * */
+mible_status_t mible_gatts_notify_or_indicate(uint16_t conn_handle,
+                                              uint16_t srv_handle, uint16_t char_value_handle, uint8_t offset,
+                                              uint8_t *p_value, uint8_t len, uint8_t type)
+{
+  bStatus_t ret = SUCCESS;
+  if (NULL == p_value)
+  {
+      return MI_ERR_INVALID_ADDR;
+  }
+
+
+  service_db_t *psrv_db = find_srv_db(srv_handle);
+  if (NULL == psrv_db)
+  {
+      return MIBLE_ERR_ATT_INVALID_ATT_HANDLE;
+  }
+
+  uint8_t attr_idx = char_value_handle - psrv_db->service.attrs[0].handle;
+  gattAttribute_t* pattr = &(psrv_db->service.attrs[attr_idx]);
+
+  /* get uuid */
+  uint8_t uuid_len = pattr->type.len;
+  uint16_t uuid = 0;
+  if (uuid_len == 2)
+  {
+    uuid = pattr->type.uuid[1];
+    uuid = (uuid << 8) | pattr->type.uuid[0];
+    if(is_fixed_uuid(uuid)){
+      MI_LOG_ERROR("gatt server set failed: fixed uuid(%d)", uuid);
+      return MIBLE_ERR_GATT_INVALID_ATT_TYPE;
+    }
+  }
+
+  gattCharCfg_t* pcccd = (gattCharCfg_t*)(pattr[1].pValue);
+  uint16_t cccd_value = 0; 
+  cccd_value = GATTServApp_ReadCharCfg(conn_handle, pcccd);
+  if(cccd_value != (uint16_t)type){
+    return MIBLE_ERR_GATT_INVALID_ATT_TYPE;
+  }
+
+  if (type == GATT_CLIENT_CFG_NOTIFY)
+  {
+    attHandleValueNoti_t* pnotify = (attHandleValueNoti_t*)osal_bm_alloc(sizeof(attHandleValueNoti_t));
+    if(pnotify == NULL)
+      return MI_ERR_NO_MEM;
+    osal_memset(pnotify, 0, sizeof(attHandleValueNoti_t));
+    pnotify->handle = char_value_handle;
+    pnotify->len = offset + len;
+    memcpy(pattr->pValue + offset, p_value, len);
+    memcpy(pnotify->value, pattr->pValue, pnotify->len);
+    ret = GATT_Notification(conn_handle, pnotify, FALSE);
+    osal_bm_free(pnotify);
+  }
+  else if(type == GATT_CLIENT_CFG_INDICATE)
+  {
+    attHandleValueInd_t* pindicate = (attHandleValueInd_t*)osal_bm_alloc(sizeof(attHandleValueInd_t));
+    if(pindicate == NULL)
+      return MI_ERR_NO_MEM;
+    osal_memset(pindicate, 0, sizeof(attHandleValueInd_t));
+    pindicate->handle = char_value_handle;
+    pindicate->len = offset + len;
+    memcpy(pattr->pValue + offset, p_value, len);
+    memcpy(pindicate->value, pattr->pValue, pindicate->len);
+    ret = GATT_Indication(conn_handle, pindicate, FALSE, get_taskID());
+    
+    osal_bm_free(pindicate);
```

Comment:
```suggestion
    ret = GATT_Indication(conn_handle, pindicate, FALSE, get_taskID());
    if (ret != SUCCESS) {
        osal_bm_free(pindicate);
    }
```

这一块有问题，TI的协议栈应该是在返回值为错时，才主动释放内存，否则应当由协议栈自动释放。

Reply:
这里协议栈有数据复制的，只是做个传递，所以用完就释放